Add 1D and 2D Wave Equation Examples using Mimetic Methods

ArshiaIlaty opened the pull request:
## 1. wave1d.cpp: Implements the 1D wave equation solver using:
   - Second-order accurate spatial discretization
   - Position Verlet time integration scheme
   - Mimetic Laplacian operator
   - Visualization using Gnuplot

## 2. wave2d.cpp: Implements the 2D wave equation solver featuring:
   - Second-order accurate spatial discretization in both dimensions
   - Position Verlet time integration
   - 2D Mimetic Laplacian operator with Robin boundary conditions
   - Interactive 3D visualization using Gnuplot

## Key Features:
- Uses existing mimetic operators (Gradient, Divergence, Laplacian)
- Implements proper boundary conditions using RobinBC
- Provides real-time visualization of wave propagation
- Maintains consistency with existing codebase style and structure
- Updated Makefile to include new examples

## Dependencies
- Armadillo (Linear algebra library)
- Boost (For filesystem and gnuplot-iostream)
- Gnuplot (For visualization)
- OpenBLAS (Optional, for better performance)
- SuperLU (For sparse matrix operations)
- Incorporate C++20 for better filesystem handling

### Installation on different platforms:

#### macOS (using Homebrew):
```
brew install armadillo
brew install boost
brew install gnuplot
brew install openblas
```
#### Ubuntu/Debian:
```
sudo apt-get install libarmadillo-dev
sudo apt-get install libboost-all-dev
sudo apt-get install gnuplot
sudo apt-get install libopenblas-dev
sudo apt-get install libsuperlu-dev
```

### PR Structure
examples/cpp/
├── wave1d.cpp          # First case of 1D wave equation
├── wave1d_case2.cpp    # Second case of 1D wave equation
├── wave2d.cpp          # First case of 2D wave equation
├── wave2d_case2.cpp    # Second case of 2D wave equation
└── Makefile           # Build system for examples

## Testing
1. Build the examples:
```
cd examples/cpp
make wave1d wave1d_case2 wave2d wave2d_case2
```
2. Run individual examples
```
./wave1d
./wave1d_case2
./wave2d
./wave2d_case2
```

![Wave1Dcpp](https://github.com/user-attachments/assets/88f3bd68-3079-45bd-93ae-b651f1cb46e4)
![Wave2Dcpp](https://github.com/user-attachments/assets/f42c3d03-d9dc-48f3-94ec-e0ce78437177)

aboada commented on examples/cpp/Makefile:
```diff
@@ -1,7 +1,7 @@
 
 ifndef ARMA
-	ARMA = /home/johnny/armadillo-10.2.1
+    ARMA = /home/johnny/armadillo-10.2.1
```
Does this work with spaces instead of tabs?

aboada commented on examples/cpp/wave1D_case2.cpp:
```diff
@@ -0,0 +1,107 @@
+        std::ofstream outfile("solution_" + std::to_string(step) + ".dat");
+        for (int i = 0; i < m+2; i++) {
+            outfile << xgrid(i) << " " << u(i) << std::endl;
+        }
```
You should close the file at this point. That is, `outfile.close();`.

ArshiaIlaty replied:
std::ofstream will automatically close the file when it goes out of scope. However, you are totally right. I should explicitly close the file.

aboada replied:
You are right. The destructor of `std::ofstream` will close the file when it is out of scope. The advantage is that you can check for errors during closing (which is probably overkill for this code).

In short, no need to explicitly close the file.

aboada commented on examples/cpp/wave1D_case2.cpp:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        // Save data for plotting
+        std::ofstream outfile("solution_" + std::to_string(step) + ".dat");
```
You can use `std::stringstream` here for faster string-building (vs `std::string`).
```
std::stringstream filename;
filename << "solution_" << step << ".dat";
std::ofstream outfile(filename.str());```

ArshiaIlaty replied:
Thanks for the suggestion. Using std::stringstream is indeed more efficient

aboada commented on examples/cpp/wave1D_case2.cpp:
```diff
@@ -0,0 +1,107 @@
+    // Animation loop
+    for (int step = 0; step <= n_steps; step++) {
+        gp << "set title '1D Wave equation t = " << step*dt << "'\n";
+        gp << "plot 'solution_" << step << ".dat' with lines linestyle 1 notitle\n";
```
Does `replot` make sense here?
http://www.gnuplot.info/docs_4.2/node150.html

ArshiaIlaty replied:
In animation, I want to create a new plot for each frame, so I am using different data files for each frame. Will read the documentation to make it more optimized.

aboada commented on examples/cpp/wave1D_case2.cpp:
```diff
@@ -0,0 +1,107 @@
+    // Cleanup solution files
+    for (int step = 0; step <= n_steps; step++) {
+        std::string filename = "solution_" + std::to_string(step) + ".dat";
+        std::remove(filename.c_str());
```
Since you added support for C++17, you should use std::filesystem::remove here.
See https://en.cppreference.com/w/cpp/filesystem/remove.

aboada reviewed (commented):
I only reviewed `examples/cpp/wave1D_case2.cpp`. My comments are also applicable to the other _cpp_ files you are adding in this PR. 
Let me know if you want me to review the other files.
Thank you.

ArshiaIlaty commented on examples/cpp/Makefile:
```diff
@@ -61,5 +59,17 @@ convection_diffusion: convection_diffusion.cpp
+
 clean:
-	rm -f transport1D schrodinger1D parabolic1D elliptic1D elliptic2D convection_diffusion RK2
+	rm -f transport1D schrodinger1D parabolic1D elliptic1D elliptic2D convection_diffusion RK2 wave1d wave1D_case2 wave2d wave2D_case2
```
I have tried to stick to the structure but will modify the makefile exactly as you wrote. Thanks a lot.

ArshiaIlaty commented on examples/cpp/Makefile:
```diff
@@ -11,11 +11,11 @@ CXXFLAGS = -O0 -g
 INCPATH = -I. $(if $(ARMA), -I$(ARMA)/include) -I$(MOLE) 
 
-LIBS = -L$(MOLE) -lmole -lopenblas 
+LIBS = -L$(MOLE) -lmole -lopenblas -lboost_iostreams -lboost_system -lboost_filesystem
```
The Boost libraries are used through gnuplot-iostream.h for plotting. Specifically:

boost_iostreams: Used for pipe communication with gnuplot
boost_system: Required by boost_iostreams for system-level operations
boost_filesystem: Used for file operations in gnuplot-iostream

aboada replied:
I see. How are you getting access to those boost libraries at this level? Are there any steps required for setup? If so, are they documented?

jbrzensk replied:
Without extra information, the cmake build process fails.
```
/examples/cpp/wave1D_case2.cpp:7:10: fatal error: gnuplot-iostream.h: No such file or directory
    7 | #include <gnuplot-iostream.h>
```

ArshiaIlaty replied:
All the code sections should be added to CMakeLists.txt file
## Find required packages
```
find_package(Boost REQUIRED COMPONENTS iostreams system filesystem)
find_package(Gnuplot REQUIRED)
```

## Download gnuplot-iostream header if not exists
```
if(NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/examples/cpp/gnuplot-iostream.h")
    file(DOWNLOAD
        "https://raw.githubusercontent.com/dstahlke/gnuplot-iostream/master/gnuplot-iostream.h"
        "${CMAKE_CURRENT_SOURCE_DIR}/examples/cpp/gnuplot-iostream.h"
        SHOW_PROGRESS
    )
endif()
```

## Add wave equation examples
```
add_executable(wave1d_case2 examples/cpp/wave1d_case2.cpp)
target_link_libraries(wave1d_case2 PRIVATE
    Boost::iostreams
    Boost::system
    Boost::filesystem
)
```

ArshiaIlaty commented on examples/cpp/Makefile:
```diff
@@ -11,11 +11,11 @@ CXXFLAGS = -O0 -g
 CXXFLAGS = -O3
 endif
-CXXFLAGS += -fopenmp -DARMA_DONT_USE_WRAPPER 
+CXXFLAGS += -fopenmp -DARMA_DONT_USE_WRAPPER -std=c++17
```
Would you please clarify? What is a changeset? Do you mean the READme file?

aboada replied:
I'm referring to the description of your changes. Basically, add C++17 when you mention dependencies in:

> Dependencies:
> 
> Armadillo for linear algebra operations
> Boost for file operations and visualization
> Gnuplot for real-time plotting

ArshiaIlaty commented on examples/cpp/wave1D_case2.cpp:
```diff
@@ -0,0 +1,107 @@
+
+int main() {
+    // Parameters
+    const int k = 4;           // Order of accuracy (spatial)
```
good suggestion. I used the same naming as the matlab file. However, sure, Why not. I will name them better

ArshiaIlaty commented on examples/cpp/wave1D_case2.cpp:
```diff
@@ -0,0 +1,107 @@
+    const double c = 100.0;    // Wave speed
+    const double dt = dx/(2*c);// Time step (CFL condition)
+    const double T = 0.06;     // Total simulation time
+    const int n_steps = static_cast<int>(T/dt);
```
I will change that step to: constexpr int n_steps = static_cast<int>(std::floor(T/dt));

Also, I can do: 
constexpr int n_steps = static_cast<int>(std::ceil(T/dt));
or
constexpr int n_steps = static_cast<int>(std::round(T/dt));
Which one do you suggest?

aboada replied:
I'm thinking that it doesn't hurt if you run an extra step, right? If so, `std::ceil`. But I feel that any of those 3 will do.

ArshiaIlaty replied:
I did some research on this topic, and this what I found: In C++17, std::floor and std::pow are not marked as constexpr functions in the standard library. Result: In C++20, many math functions including std::pow are constexpr. Should I use the version instead of 17? Do you have any suggestions?

ArshiaIlaty replied:
I can keep them as const:
`
const int kNumSteps = static_cast<int>(std::floor(kTotalTime/kDt));

const double kTheta = 1.0/(2.0-std::pow(2.0, 1.0/3.0));
`

or use their precalculated value:
kNumSteps = 81 (from kTotalTime/kDt)
kTheta = 1.351207191959657 (from 1.0/(2.0-std::pow(2.0, 1.0/3.0)))

ArshiaIlaty commented on examples/cpp/wave1D_case2.cpp:
```diff
@@ -0,0 +1,107 @@
+
+    // Create Laplacian operator (k=4 for 4th order accuracy)
+    Laplacian L(k, m, dx);
+    const arma::sp_mat& L_sp = static_cast<const arma::sp_mat&>(L);
```
Indeed. It was unnecessary. Will modify the code and use the L directly.

ArshiaIlaty commented:
Would you please review the new version of the wave1d_case2? I have tried to apply all your suggestions. Please let me know if I need to do anything else. Thanks a ton.

aboada commented on examples/cpp/wave1D_case2.cpp:
```diff
@@ -0,0 +1,111 @@
+        std::stringstream filename;
+        filename << "solution_" << step << ".dat";
+        std::error_code ec;
+        std::filesystem::remove(filename.str(), ec);
```
I think you can check the error code directly:
```
if(!std::filesystem::remove(filename.str(), ec)) {
... // print error message.
}

ArshiaIlaty replied:
I do have a question. I cannot make this plot interactive, and whenever I run the code, I should wait until the end of the plotting. However, it is better to give an option to the user to close the window anytime. I cannot find any solution for that. Do you have any recommendations?

ArshiaIlaty replied:
Is this what you mean?
```
        if (!std::filesystem::remove(filename.str(), ec)) {
            std::cerr << "Error deleting file " << filename.str() << ": " << ec.message() << std::endl;
        }
```
Thanks for the comment

aboada commented on examples/cpp/wave1D_case2.cpp:
```diff
@@ -0,0 +1,111 @@
+
+// Force function (equivalent to MATLAB's F = @(x) (c^2)*L*x)
+arma::vec calculateForce(const Laplacian& L, const arma::vec& x, const double c) {
+    return (c * c) * (static_cast<const arma::sp_mat&>(L) * x);
```
I don't think you need this casting, right?
`static_cast<const arma::sp_mat&>(L)`, should be just `L`

aboada replied:
Just noticed, you are recomputing `c^2` every time you call this function. `calculateForce()` is called several times in the time integration loop. Pre-compute `c^2`.

aboada replied:
If performance is not an issue for this example, don't worry too much about extracting all the other constants from the loop.

ArshiaIlaty replied:
I converted the function to this:
```
arma::vec calculateForce(const Laplacian& L, const arma::vec& x, const double c_squared) {
    // Removed unnecessary casting since Laplacian inherits from sp_mat
    return c_squared * (L * x);
}
```
However, I faced with this issue:
wave1d_case2.cpp:15:27: error: invalid operands to binary expression ('const Laplacian' and 'const arma::vec' (aka 'const Col<double>'))
   15 |     return c_squared * (L * x);

So, it seems that even though Laplacian inherits from sp_mat, we still need to cast it explicitly in this case. Do you have any suggestions in this regard?

aboada replied:
You are only including `#include "laplacian.h"`. The operation overload for Lap-vec multiplication is implemented in `https://github.com/csrc-sdsu/mole/blob/master/src/cpp/operators.h`. 

You can try to include `operators.h`, or remove `laplacian.h` and include `mole.h` as in `examples/cpp/transport1D.cpp`.

I think including `mole.h` is a practice that should be discouraged as you want to include only the headers that are strictly necessary for the code to compile and run correctly. However, the way the library is designed and implemented skews you towards including it.

Hopefully it works. If it doesn't, I should have some time this weekend to test your changes and be of more help to you.

aboada commented:
> Would you please review the new version of the wave1d_case2? I have tried to apply all your suggestions. Please let me know if I need to do anything else. Thanks a ton.

It looks great, other than the two minor things I mentioned. Thanks for the changes!

aboada commented:
> ## 1. wave1d.cpp: Implements the 1D wave equation solver using:
> * Second-order accurate spatial discretization
> * Position Verlet time integration scheme
> * Mimetic Laplacian operator
> * Visualization using Gnuplot
> 
> ## 2. wave2d.cpp: Implements the 2D wave equation solver featuring:
> * Second-order accurate spatial discretization in both dimensions
> * Position Verlet time integration
> * 2D Mimetic Laplacian operator with Robin boundary conditions
> * Interactive 3D visualization using Gnuplot
> 
> ## Key Features:
> * Uses existing mimetic operators (Gradient, Divergence, Laplacian)
> * Implements proper boundary conditions using RobinBC
> * Provides real-time visualization of wave propagation
> * Maintains consistency with existing codebase style and structure
> * Updated Makefile to include new examples
> 
> ## Dependencies
> * Armadillo (Linear algebra library)
> * Boost (For filesystem and gnuplot-iostream)
> * Gnuplot (For visualization)
> * OpenBLAS (Optional, for better performance)
> * SuperLU (For sparse matrix operations)
> * Incorporate C++20 for better filesystem handling
> 
> ### Installation on different platforms:
> #### macOS (using Homebrew):
> ```
> brew install armadillo
> brew install boost
> brew install gnuplot
> brew install openblas
> ```
> 
> #### Ubuntu/Debian:
> ```
> sudo apt-get install libarmadillo-dev
> sudo apt-get install libboost-all-dev
> sudo apt-get install gnuplot
> sudo apt-get install libopenblas-dev
> sudo apt-get install libsuperlu-dev
> ```
> 
> ### PR Structure
> examples/cpp/ ├── wave1d.cpp # First case of 1D wave equation ├── wave1d_case2.cpp # Second case of 1D wave equation ├── wave2d.cpp # First case of 2D wave equation ├── wave2d_case2.cpp # Second case of 2D wave equation └── Makefile # Build system for examples
> 
> ## Testing
> 1. Build the examples:
> 
> ```
> cd examples/cpp
> make wave1d wave1d_case2 wave2d wave2d_case2
> ```
> 
> 2. Run individual examples
> 
> ```
> ./wave1d
> ./wave1d_case2
> ./wave2d
> ./wave2d_case2
> ```
> 
> ![Wave1Dcpp](https://private-user-images.githubusercontent.com/58371460/410635547-88f3bd68-3079-45bd-93ae-b651f1cb46e4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzNzQ2NDgsIm5iZiI6MTczOTM3NDM0OCwicGF0aCI6Ii81ODM3MTQ2MC80MTA2MzU1NDctODhmM2JkNjgtMzA3OS00NWJkLTkzYWUtYjY1MWYxY2I0NmU0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEyVDE1MzIyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIwYTE5YmIwY2JkODdiZmZjMThkMWNmMjM1MjA5ZTdkNjMwZTQ2YjczYTcwZjZkNmU4MDUxNjFlNjNiNWNmNTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.eYxDzIloHzXT-emqXYZmqgJTfiAO_6NDIuNXHniKd7s) ![Wave2Dcpp](https://private-user-images.githubusercontent.com/58371460/410636196-f42c3d03-d9dc-48f3-94ec-e0ce78437177.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzNzQ2NDgsIm5iZiI6MTczOTM3NDM0OCwicGF0aCI6Ii81ODM3MTQ2MC80MTA2MzYxOTYtZjQyYzNkMDMtZDlkYy00OGYzLTk0ZWMtZTBjZTc4NDM3MTc3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEyVDE1MzIyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJlZDQxNTM5MGQ0ZTY1NDI3MGNkNWFkNzRhZGQ3MjUwMzdjNGE4NjI0ZjdhYTQzNGRjN2M2ZmQ1NTkwZjczNGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.z8N7O8MNsO2tqVUXYQFaNJ2QuQlvTD2PHinxDRYnpyg)

This information looks great! Would it be possible to have it on a readme file?

ArshiaIlaty commented:
I updated the README file and modified the wave2d code as well. Please review wave2d, wave1d, and wave1d_case2. Let me know if you have any questions. If everything looks good, I will proceed with updating wave2d_case2 as well.

ArshiaIlaty commented:
This is the content for CMakeLists.txt if it helps for code reproducibility.
```
# examples_C++ Configuration
include_directories("${CMAKE_SOURCE_DIR}/src/cpp")

# Find all .cpp files in the examples directory
file(GLOB EXAMPLE_SOURCES *.cpp)

# Create executables for each source file
foreach(EXAMPLE_SOURCE ${EXAMPLE_SOURCES})
    get_filename_component(EXAMPLE_NAME ${EXAMPLE_SOURCE} NAME_WE)
    add_executable(${EXAMPLE_NAME} ${EXAMPLE_SOURCE})
    target_link_libraries(${EXAMPLE_NAME} PUBLIC mole_C++ ${LINK_LIBS})
endforeach()

# Find required packages
find_package(Boost REQUIRED COMPONENTS iostreams system filesystem)
find_package(Gnuplot REQUIRED)

# Download gnuplot-iostream header if not exists
if(NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/examples/cpp/gnuplot-iostream.h")
    file(DOWNLOAD
        "https://raw.githubusercontent.com/dstahlke/gnuplot-iostream/master/gnuplot-iostream.h"
        "${CMAKE_CURRENT_SOURCE_DIR}/examples/cpp/gnuplot-iostream.h"
        SHOW_PROGRESS
    )
endif()

# Add wave equation examples
add_executable(wave1d_case2 examples/cpp/wave1d_case2.cpp)
target_link_libraries(wave1d_case2 PRIVATE
    Boost::iostreams
    Boost::system
    Boost::filesystem
)
```

jbrzensk commented:
This still does not compile.
```
wave2D_case2.cpp:7:10: fatal error: gnuplot-iostream.h: No such file or directory
```
If the cmake file needs to be edited, then it should be included in this PR.

ArshiaIlaty commented:
I updated the cmakelists and makefile. Please try to run it again with all the packages required. I proposed some new packages to be installed in the PR.

jbrzensk commented:
New error on using cmake.

```
-- Found Gnuplot: /usr/bin/gnuplot (found version "6.0.0") 
-- Configuring done (31.8s)
CMake Error at examples/cpp/CMakeLists.txt:28 (add_executable):
  Cannot find source file:

    examples/cpp/wave1d_case2.cpp

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm
  .ccm .cxxm .c++m .h .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90
  .f95 .f03 .hip .ispc


CMake Error at examples/cpp/CMakeLists.txt:28 (add_executable):
  No SOURCES given to target: wave1d_case2

```

ArshiaIlaty commented:
Please try it again. The file name was wave1D_case2.cpp in my repo. It should have been wave1d_case2.cpp. I modified the names, and it should work now.

JananiPSrinivasan commented:
Hi @ArshiaIlaty, Please try working with the CMake file I have mentioned below. Replace your Cmake file with this file as I have restructured it and avoided redundancy. Hope it resolves your issue.

    # Include directories
    include_directories("${CMAKE_SOURCE_DIR}/src/cpp")

    # Find all .cpp files in the examples directory
    file(GLOB EXAMPLE_SOURCES examples/cpp/*.cpp)

    # Find required packages
    find_package(Boost REQUIRED COMPONENTS iostreams system filesystem)
    find_package(Gnuplot REQUIRED)

    # Download gnuplot-iostream header if not exists
    if(NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/examples/cpp/gnuplot-iostream.h")
        file(DOWNLOAD
            "https://raw.githubusercontent.com/dstahlke/gnuplot-iostream/master/gnuplot-iostream.h"
            "${CMAKE_CURRENT_SOURCE_DIR}/examples/cpp/gnuplot-iostream.h"
            SHOW_PROGRESS
        )
    endif()

    # Create executables and link them appropriately
    foreach(EXAMPLE_SOURCE ${EXAMPLE_SOURCES})
        get_filename_component(EXAMPLE_NAME ${EXAMPLE_SOURCE} NAME_WE)
        add_executable(${EXAMPLE_NAME} ${EXAMPLE_SOURCE})
        target_link_libraries(${EXAMPLE_NAME} PUBLIC mole_C++ ${LINK_LIBS})

        # Special handling for wave1D_case2
        if(EXAMPLE_NAME STREQUAL "wave1D_case2")
            target_link_libraries(${EXAMPLE_NAME} PRIVATE Boost::iostreams Boost::system Boost::filesystem)
        endif()
    endforeach()

Thanks!

ArshiaIlaty commented:
Hi @JananiPSrinivasan 
Thank you so much for cleaning the file. I appreciate your help and effort. Please let me know if there is anything else I should do. I think all the code should run without any problems on your end.

jbrzensk commented:
Hi @ArshiaIlaty,

The other maintainers and I have been chatting about your examples. They are pretty cool with the interactive visualization. Surely, this is something the other examples do not do. We are worried about the additional requirements your examples put on the MOLE library.

The library already has a set of dependencies, and most are nonnegotiable, having to do with the mathematics involved and solving of sparse systems, which the library is primarily designed to do. With that, we are very hesitant to add additional dependencies, namely the boost library, the gnuplot library, and the C++20 requirement.

Boost is a great library but is unnecessary for the core functionality of the MOLE library. Same for gnuplot. Gnuplot ships standard with Linux and interacts natively with C++, but requiring the dev libraries also might be asking too much of the casual user. 

And finally, the C++20 requirement. We wrote most of the library to be as compatible with older software as possible. Some of the codes using MOLE are older, with their own even older dependencies, and forcing a requirement to use 20 may have unforeseen implications on these other applications.

What I do think is that you should keep the final working examples. A good compromise may be to simplify your examples using gnuplot for the MOLE library, and then you have your own hosted interactive examples, and we could link to your code from the MOLE example or repositories, something like a comment in wave1d.cpp like "If you want a more interactive Wave1D, go to https://github.com/ArshiaIlaty/mole_Ash and check out the interactive version".

I would like @aboada , @jcorbino, @cpaolini, @valeriabarra  to chime in on this in case I am out of line. And I am sorry @ArshiaIlaty for taking so long to bring this up.

mdumett commented:
Please, even though that is not in the comments of the MATLAB code, indicate in the top comments of your code which equation is being solved, its time and space domains and its initial and boundary conditions.

mdumett commented:
On 2/13, @jbrzensk suggested removing in your code any dependencies on gnuplot dev, C++ 20 and boost. Were you able to do that?

ArshiaIlaty commented:
I have been working on it and will push the code for your review. Thank you.

ArshiaIlaty commented:
@aboada @jbrzensk Hi, hope you are doing well. Please share your thoughts on the new version of these four examples. I guess the conflicts for the gitignore and Makefile should be solved on your end. Please let me know if I should modify it. Thank you, Jared, for your comments, and thank you, Angel, for reviewing my code and guidance.

mdumett commented:
Which version of the C++ compiler does your code requires?

ArshiaIlaty commented:
> Which version of the C++ compiler does your code requires?

No need for version 20. I utilized 14 in the Makefile

ArshiaIlaty commented:
> > @aboada @jbrzensk Hi, hope you are doing well. Please share your thoughts on the new version of these four examples. I guess the conflicts for the gitignore and Makefile should be solved on your end. Please let me know if I should modify it. Thank you, Jared, for your comments, and thank you, Angel, for reviewing my code and guidance.
> 
> Thanks, let me know when your PR is ready for review.

It is ready to review. I made the visualization cleaner. Please let me know your thoughts on the conflict of the .gitignore and Makefile. I assume I can delete my version or if you want I can resolve the conflict. Thank you

mdumett commented:
@aboada: Is it required that the code is in C++ 11 or earlier? Ash's code seems to be C++ 14.

aboada commented:
> @aboada: Is it required that the code is in C++ 11 or earlier? Ash's code seems to be C++ 14.

Don't think we have established that dependency. Ash's code compiles without errors on my fresh Ubuntu installation (using the `build-essential` package).

From what I can see, my build is configured to use C++14, see: 
```
mando@karak-kadrin:~/src/mole$ grep CMAKE_CXX -IrE | head -5
CMakeLists.txt:set(CMAKE_CXX_STANDARD 14)
CMakeLists.txt:set(CMAKE_CXX_STANDARD_REQUIRED True)
CMakeLists.txt:message(STATUS "Detected CXX Compiler ID: ${CMAKE_CXX_COMPILER_ID}")
CMakeLists.txt:if (CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang")
CMakeLists.txt:    set(CMAKE_CXX_FLAGS "-O3 -Xclang -fopenmp -DARMA_DONT_USE_WRAPPER -DARMA_USE_SUPERLU")```

ArshiaIlaty commented:
> @aboada: Is it required that the code is in C++ 11 or earlier? Ash's code seems to be C++ 14.

Even if there is a required dependency, I assume my code runs with version 11 as well. I have not tried to run it with 11, but I can give it a shot.

jbrzensk commented on examples/cpp/wave2d.cpp:
```diff
@@ -0,0 +1,143 @@
+        }
+    }
+    arma::vec u = arma::vectorise(U_init);
+    arma::vec v = arma::zeros<arma::vec>(I_scaled.n_rows);
```
```cpp
arma::vec v( I_scaled.rows, fill::zeros)
```

jbrzensk commented on examples/cpp/wave2d_case2.cpp:
```diff
@@ -0,0 +1,146 @@
+    Interpol I2(true, kNumCellsX, kNumCellsY, 0.5, 0.5);
+
+    // Cast and combine operators
+    const arma::sp_mat& L_sp = static_cast<const arma::sp_mat&>(L);
```
Already sparse.

jbrzensk commented on examples/cpp/wave2d_case2.cpp:
```diff
@@ -0,0 +1,146 @@
+    const arma::sp_mat& L_sp = static_cast<const arma::sp_mat&>(L);
+    const arma::sp_mat& BC_sp = static_cast<const arma::sp_mat&>(BC);
+    arma::sp_mat combined = L_sp + BC_sp;
+    arma::sp_mat I_scaled = kDt * static_cast<arma::sp_mat>(I);
```
If you do the fix from wave2d, this doesn't need the casting

jbrzensk commented on examples/cpp/wave2d_case2.cpp:
```diff
@@ -0,0 +1,146 @@
+        }
+    }
+    arma::vec u = arma::vectorise(U_init);
+    arma::vec v = arma::zeros<arma::vec>(I_scaled.n_rows);
```
```cpp
arma::vec v( I_scaled.rows, fill::zeros)
```

jbrzensk commented on examples/cpp/wave2d_case2.cpp:
```diff
@@ -0,0 +1,146 @@
+        return EXIT_FAILURE;
+    }
+
+    plot_script << "set title 'Wave Equation Solution (2D Case 2)'\n";
```
Still a nice plot!

jbrzensk reviewed (commented):
This is nice. I like the plotting, so much simpler!
You can remove a bunch of files, all that is needed is the wave1d, wave1d_case2, wave2d, wave2d_case2.

Most of my comments revolve around the casting MOLE operators to sparse matrices. All MOLE operators are sparse matrices by design. And there is a fix for the Interpolator error that shows up if you don't cast to sparse.
Please edit the src/cpp/operators.h file and add that to the pull request.

ArshiaIlaty commented:
@jbrzensk Hi Jared,

Thanks for reveiwing the code. I have tried to address all of your comments. Please recheck and let me know if there is something else I need to do.

jbrzensk commented:
Cool beans!
Looks like you need to fetch from MOLE to catch up on the last 175 commits you are missing!
![image](https://github.com/user-attachments/assets/0dc06e48-2e1f-413c-afe6-feb27a0e0d95)

You can Google "fetch from upstream" for more info. Don't blindly follow this, but you will probably:
```
git remote add upstream https://github.com/csrc-sdsu/mole.git
git fetch upstream
git merge upstream master
git push origin master
```
Or something like that. It may generate some issues.

ArshiaIlaty commented:
@jbrzensk Hi Jared, I did have a lot of conflicts and needed to rebase and so on. I think the final version of the codes and operators.h  have been pushed. Please let me know if I should do anything else.

ArshiaIlaty commented:
<img width="1487" alt="Screenshot 2025-05-07 at 5 03 02 PM" src="https://github.com/user-attachments/assets/2ae7a664-b1f4-49e0-a399-5c315b124b5b" />

I think I still have some issues with the operators.h implementation. Am I right?

ArshiaIlaty commented:
@jbrzensk Thank you Jared for all of your guidance.
